CMakeLists.txt

```diff
@@ -174,10 +176,12 @@ endif()
 unset(OLD_REQUIRED_LIBRARIES)
 
 # Include liblog for Android logging
-check_library_exists(log __android_log_print "" HAVE_LIBLOG)
-if(HAVE_LIBLOG)
-    set(EXTRA_LIBS log ${EXTRA_LIBS})
-    set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} log)
+if(ANDROID)
+    check_library_exists(log __android_log_print "" HAVE_LIBLOG)
+    if(HAVE_LIBLOG)
+        set(EXTRA_LIBS log ${EXTRA_LIBS})
+        set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} log)
+    endif()
```

Comment:
Just out of curiosity, how is this change related to a framework creation?

Reply:
please refer to: https://github.com/kcat/openal-soft/pull/452

Reply:
Still not sure how this Android-specific thing affects creation of a framework.

Reply:
it will lead link issue, -llog not exist

Reply:
`HAVE_LIBLOG` should not be defined after `check_library_exists()` call if there is no such library.

Reply:
but it doesn't work, you can try my fork on macOS 11.15, xcode11, cmake-3.17+

Reply:
Please define "doesn't work". If it was detection errors, we would like to know why did it happen.

Regardless of the actual issue, I cannot comprehend why this affects framework creation.
I built static and dynamic libraries with various versions of Xcode, and never had any issues with this.

Reply:
Please look into this, too. If `-llog` doesn't exist, it should not try to use it and so won't be a link problem. It could only succeed if the function `__android_log_print` specifically exists along with `liblog`. Specifically checking for Android first would be redundant. Is the check succeeding for some reason? What does the cmake log (in `CMakeFiles/CMakeOutput.log`) say for the function check?

Reply:
See the snapshot:
![image](https://user-images.githubusercontent.com/6977319/90760970-3f0f3100-e315-11ea-9d38-558a1afc4b6f.png)

Reply:
`check_library_exists` caches its results, so if `HAVE_LIBLOG` was already set to true for some reason, it'll forever stay true and not be re-checked until the cache is cleared. What does `CMakeOutput.log` show for the `__android_log_print` check?

Reply:
Yeah, I clear the cache, I will upload two CMakeOutput.log later
1. CMake without toolchain ```ios.toolchain.cmake``` use iphone simulator sdk, the check should be fine, log not exist on ios
2. CMake with toolchain ```ios.toolchain.cmake`` use iphone device sdk, the check_library_exists not fine, will ```HAVE_LIBLOG``` will be ```true```

Reply:
See the attachment file, it contains 2 file of CMake output for simulator_x64 and device_arm:

[CMakeOutputs.zip](https://github.com/kcat/openal-soft/files/5106631/CMakeOutputs.zip)

Reply:
Generate Command
* device arm: cmake -S . -GXcode -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_REQUIRE_COREAUDIO=ON  -DALSOFT_EMBED_HRTF_DATA=YES -DLIBTYPE=SHARED -DALSOFT_UTILS=OFF -DALSOFT_EXAMPLES=OFF -DALSOFT_INSTALL=OFF -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_OSX_FRAMEWORK=ON -DCMAKE_TOOLCHAIN_FILE=cmake/ios.toolchain.cmake -Bbuild_arm
* iphone simulator: cmake -S . -GXcode -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_REQUIRE_COREAUDIO=ON  -DALSOFT_EMBED_HRTF_DATA=YES -DLIBTYPE=SHARED -DALSOFT_UTILS=OFF -DALSOFT_EXAMPLES=OFF -DALSOFT_INSTALL=OFF -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_OSX_FRAMEWORK=ON -DCMAKE_TOOLCHAIN_FILE=cmake/ios.toolchain.cmake -Bbuild_arm

Reply:
> Yeah, I clear the cache, I will upload two CMakeOutput.log later
> 
> 1. CMake without toolchain `ios.toolchain.cmake` use iphone simulator sdk, the check should be fine, log not exist on ios
> 2. CMake with toolchain `ios.toolchain.cmake` use iphone device sdk, the check_library_exists not fine, will  `HAVE_LIBLOG` will be `true`

I notice all `check_library_exists` checks seem to succeed; the others, libm, libdl, and librt, make sense to on a POSIX-based system, but liblog shouldn't. It looks like this may be at fault:
> ENABLE_STRICT_TRY_COMPILE: (1|0) Enables or disables strict try_compile() on all Check* directives (will run linker
>   to actually check if linking is possible). **Default 0** (false, will set CMAKE_TRY_COMPILE_TARGET_TYPE to STATIC_LIBRARY)

I don't know why it defaults to false, since a static library avoids any link-time error for missing symbols and dependent libraries, making checks for symbols and libraries like this succeed despite one (or both) not existing. It also seems to be a pretty useless option, since it just sets a variable that the user can set; rather than setting `ENABLE_STRICT_TRY_COMPILE=0` or `1`, set `CMAKE_TRY_COMPILE_TARGET_TYPE=STATIC_LIBRARY` or `EXECUTABLE`. The only difference is, the toolchain sets a different (and more error-prone) default on you.

So now there's two options. Tell all users they need to pass `-DENABLE_STRICT_TRY_COMPILE=1` to cmake when building for iPhone, or modify our local `ios.toolchain.cmake` to use a different default (or really, remove the option altogether because it can't not be set), which means we need to make sure to change it again whenever we update our local copy.

Reply:
Hi, kcat
  I want to tell you, when I pass ```-DENABLE_STRICT_TRY_COMPILE=1```, everything check failed or not found

Reply:
News: I install cmake-3.18.2, the check issue of cmake seems solved, I can confirm, previous cmake-3.18.1 doesn't works, then when use cmake-3.18.2, doesn't needs ```ios.toolchain.cmake```, the follow command can be execute succeed:
```
cmake -S . -GXcode -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_REQUIRE_COREAUDIO=ON -DALSOFT_EMBED_HRTF_DATA=YES -DLIBTYPE=SHARED -DALSOFT_UTILS=OFF -DALSOFT_EXAMPLES=OFF -DALSOFT_INSTALL=OFF -DCMAKE_CXX_EXTENSIONS=OFF -DALSOFT_OSX_FRAMEWORK=ON -DCMAKE_SYSTEM_NAME=iOS -Bbuild_arm
```
The cmake-3.18.2 released 16 hours ago

Reply:
> I want to tell you, when I pass `-DENABLE_STRICT_TRY_COMPILE=1`, everything check failed or not found

Can you provide `CMakefiles/CMakeError.log` from those failures?

> News: I install cmake-3.18.2, the check issue of cmake seems solved
> ...
> The cmake-3.18.2 released 16 hours ago

Way too new to depend on. iPhone shouldn't require a bleeding edge CMake to build for.

Reply:
OK, let me install old version and try again

Reply:
please see the attachment file:
[CMakeErrorAndOutput.zip](https://github.com/kcat/openal-soft/files/5107250/CMakeErrorAndOutput.zip)